File: playwright-typescript/src/pages/reviews-page.ts
import assert from "node:assert/strict";
import { expect, type Page } from "@playwright/test";
import { settings } from "../support/settings.js";

export class ReviewsPage {
  constructor(private readonly page: Page) {}

  open(productId: number): Promise<unknown> { return this.page.goto(`${settings.baseUrl}/products/${productId}/view`); }
  expectFormHidden(): Promise<void> { return expect(this.page.locator("form.review-form")).toHaveCount(0); }

  async expectSignInLink(): Promise<void> {
    const link = this.page.locator("#reviews").getByRole("link", { name: "Sign in", exact: true });
    await expect(link).toBeVisible();
    await expect(link).toHaveAttribute("href", `/signin?next=${this.page.url().replace(settings.baseUrl, "")}`);
  }

  async expectFields(): Promise<void> {
    await expect(this.page.getByRole("radiogroup", { name: "Select a rating" })).toBeVisible();
    await expect(this.page.getByLabel("Feedback")).toBeVisible();
  }

  expectSubmit(): Promise<void> { return expect(this.page.getByRole("button", { name: "Submit review", exact: true })).toBeVisible(); }

  async submitFeedbackWithoutRating(feedback: string): Promise<void> {
    await this.page.getByLabel("Feedback").fill(feedback);
    await this.submit();
  }

  expectRatingRequired(): Promise<void> { return expect(this.page.locator(".form-error")).toHaveText("Select a rating using the stars."); }

  async selectRating(rating: number): Promise<void> {
    const button = this.page.locator(`.star-picker-button[data-index="${rating}"]`);
    const box = await button.boundingBox();
    assert.ok(box);
    await button.click({ position: { x: box.width * 0.75, y: box.height / 2 } });
    await expect(this.page.locator("#review-rating-value")).toHaveValue(rating.toFixed(1));
  }

  submit(): Promise<void> { return this.page.getByRole("button", { name: "Submit review", exact: true }).click(); }
  submitUpdated(): Promise<void> { return this.page.getByRole("button", { name: "Update review", exact: true }).click(); }

  async expectFeedbackRequired(): Promise<void> {
    const field = this.page.getByLabel("Feedback");
    assert.equal(await field.evaluate((element: HTMLTextAreaElement) => element.validity.valueMissing), true);
    assert.ok(await field.evaluate((element: HTMLTextAreaElement) => element.validationMessage));
  }

  enterFeedback(feedback: string): Promise<void> { return this.page.getByLabel("Feedback").fill(feedback); }

  async expectReview(username: string, rating: number, comment: string): Promise<void> {
    const review = this.page.locator(".review-list li").filter({ hasText: username });
    await expect(review).toHaveCount(1);
    await expect(review).toContainText(comment);
    await expect(review.locator(".review-rating")).toHaveAttribute("data-rating", rating.toFixed(1));
    await expect(review.locator(".review-header span")).toHaveText(/^(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec) \d{1,2}, \d{4}$/);
  }

  async expectRatingSummary(rating: number, count: number): Promise<void> {
    const summary = this.page.locator(".product-detail-rating");
    await expect(summary).toContainText(`${rating.toFixed(1)} / 5`);
    await expect(summary).toContainText(`(${count} reviews)`);
  }
}

File: playwright-csharp/Pages/ReviewsPage.cs
using System.Text.RegularExpressions;
using Fictoshop.PlaywrightTests.Support;
using Microsoft.Playwright;
using NUnit.Framework;
using static Microsoft.Playwright.Assertions;

namespace Fictoshop.PlaywrightTests.Pages;

public sealed class ReviewsPage
{
    private readonly IPage _page;

    public ReviewsPage(IPage page)
    {
        _page = page;
    }

    public Task OpenAsync(int productId) => _page.GotoAsync($"{TestSettings.BaseUrl}/products/{productId}/view");
    public Task AssertFormHiddenAsync() => Expect(_page.Locator("form.review-form")).ToHaveCountAsync(0);

    public async Task AssertSignInLinkAsync()
    {
        var link = _page.Locator("#reviews").GetByRole(AriaRole.Link, new() { Name = "Sign in", Exact = true });
        await Expect(link).ToBeVisibleAsync();
        await Expect(link).ToHaveAttributeAsync("href", "/signin?next=" + _page.Url.Replace(TestSettings.BaseUrl, ""));
    }

    public async Task AssertFieldsAsync()
    {
        await Expect(_page.GetByRole(AriaRole.Radiogroup, new() { Name = "Select a rating" })).ToBeVisibleAsync();
        await Expect(_page.GetByLabel("Feedback")).ToBeVisibleAsync();
    }

    public Task AssertSubmitAsync() => Expect(_page.GetByRole(AriaRole.Button, new() { Name = "Submit review", Exact = true })).ToBeVisibleAsync();

    public async Task SubmitFeedbackWithoutRatingAsync(string feedback)
    {
        await _page.GetByLabel("Feedback").FillAsync(feedback);
        await SubmitAsync();
    }

    public Task AssertRatingRequiredAsync() => Expect(_page.Locator(".form-error")).ToHaveTextAsync("Select a rating using the stars.");

    public async Task SelectRatingAsync(int rating)
    {
        var button = _page.Locator($".star-picker-button[data-index=\"{rating}\"]");
        var box = await button.BoundingBoxAsync();
        Assert.That(box, Is.Not.Null);
        await button.ClickAsync(new LocatorClickOptions { Position = new Position { X = box!.Width * .75f, Y = box.Height / 2 } });
        await Expect(_page.Locator("#review-rating-value")).ToHaveValueAsync($"{rating:F1}");
    }

    public Task SubmitAsync() => _page.GetByRole(AriaRole.Button, new() { Name = "Submit review", Exact = true }).ClickAsync();
    public Task SubmitUpdatedAsync() => _page.GetByRole(AriaRole.Button, new() { Name = "Update review", Exact = true }).ClickAsync();

    public async Task AssertFeedbackRequiredAsync()
    {
        var field = _page.GetByLabel("Feedback");
        Assert.That(await field.EvaluateAsync<bool>("element => element.validity.valueMissing"), Is.True);
        Assert.That(await field.EvaluateAsync<string>("element => element.validationMessage"), Is.Not.Empty);
    }

    public Task EnterFeedbackAsync(string feedback) => _page.GetByLabel("Feedback").FillAsync(feedback);

    public async Task AssertReviewAsync(string username, int rating, string comment)
    {
        var review = _page.Locator(".review-list li").Filter(new() { HasText = username });
        await Expect(review).ToHaveCountAsync(1);
        await Expect(review).ToContainTextAsync(comment);
        await Expect(review.Locator(".review-rating")).ToHaveAttributeAsync("data-rating", $"{rating:F1}");
        await Expect(review.Locator(".review-header span")).ToHaveTextAsync(new Regex(@"^(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec) \d{1,2}, \d{4}$"));
    }

    public async Task AssertRatingSummaryAsync(int rating, int count)
    {
        var summary = _page.Locator(".product-detail-rating");
        await Expect(summary).ToContainTextAsync($"{rating:F1} / 5");
        await Expect(summary).ToContainTextAsync($"({count} reviews)");
    }
}
